### ISSUE TITLE
Native support for TGs test environment

### ISSUE THREAD

**Issue by Bibo-Joshi** [python-telegram-bot/python-telegram-bot#3355]
By now, TGs [dedicated test environment](https://core.telegram.org/bots/features#dedicated-test-environment) is more prominently documented. Adding native support for it in PTB could be a desribale feature.

## ⚠️  Current status

This feature is currently not on any roadmap, since it doesn't seem to be widely used. If you would like for PTB to have this feature, please react with a 👍 to the comment. If we see a large number of requests, this issue will likely get put on a roadmap.

Edit: One can trick PTB into using the test env by setting `token='<token>/test'`. But this hack is not guaranteed to keep working in future versions of PTB …

Count of thumbs up over time, status 2024-12-31:

![Image](https://github.com/user-attachments/assets/8b48070c-6379-4330-a336-9d4a351be2a0)

<details><summary>For future reference: GraphQL query to get the data</summary>

```gql
{
  repository(owner: "python-telegram-bot", name: "python-telegram-bot") {
    issue(number: 3355) {
      reactions(first: 100) {
        nodes{
          content
          createdAt
          user {
            login
          }
        }
      }
    }
  }
}
```

</details>

## Open Questions

We should check how the url returned by `get_file` behave, i.e. if they are still `https://api.telegram.org/file/bot<Token>/…` or if they are e.g. `https://api.telegram.org/file/bot<token>/test/…`.

## Implementation options

Just quickly listing two options that come to my mind

### A `test_env` flag

That would work via `ApplicationBuilder().token(…).test_env(True)` or `Bot(…, test_env=True)` . This would be non-breaking. However, if may not play nicely with custom `base_file_url` depending on what we find out about file urls.

### Allow custom insertion of token into `base(_file)_url`

That could either work by requesting the user to pass something like `base_url = 'api.telegram.org/bot{token}/'` (this would be a breaking change) or by allowing the user to pass a `Callable[[str], str]` that accepts the token and returns the (file_)url (this could be implemented in a non-breaking manner)

**Comment by Nickelza** [comment]
It's currently the only way to test topics without having 200 members in a test group

**Comment by harshil21** [comment]
> It's currently the only way to test topics without having 200 members in a test group

As of Telegram 9.3 (Dec 30, 2022), we can now make a group of any size into forums.

**Comment by Bibo-Joshi** [comment]
For future reference: A first draft PR was already made in #3972 and if we ever decide to go ahead with this, a PR may possibly build up on that PR.

**Comment by OlegNovosad** [comment]
I've started implementing payments in telegram and this feature might be more and more requested in order to test payments in the test environment.

**Comment by Bibo-Joshi** [comment]
@OlegNovosad if you can answer the open question and ideally get TG to also document the result, I'm open to take up discussions on this again :)

**Comment by maschlr** [comment]
Using the test environment is the only way to test payments for digital goods & services (using TG stars) AFAIU. 

Testing (& receiving payments) is a very important feature for many bot projects.

@Bibo-Joshi in regards to the open question. The [docs](https://core.telegram.org/bots/features#testing-your-bot) indicate that `file_id` is unique for the different bots/environments. I read this as the url being `https://api.telegram.org/file/bot<Token>/…`

**Comment by Bibo-Joshi** [comment]
Hi @maschlr 

>  I read this as

That is the problem though ;) As validation that the url is indeed working for the test environment, I would need to see e.g. an exceprt of two curl commands, one for a `getFile` request to the test environment and one for downloading the file with the corresponding ID. If you can provide this (and as indicated above, ideally get TG to add documentation about it), I'm happey to continue the discussion.

**Comment by maschlr** [comment]
I would. It seems as if you [need an iOS device to create a TG account in the test environment](https://core.telegram.org/bots/features#creating-a-bot-in-the-test-environment) that can then be imported into the desktop app. Unfortunately, I don't have access to an iOS device :smirk_cat:

**Comment by Bibo-Joshi** [comment]
@maschlr Mh, for the TG Desktop Variant, it looks like it&#x27;s available independently of the OS ...

**Comment by Poolitzer** [comment]
@Bibo-Joshi And android beta also has test backend, but since this is not official released they don&#x27;t document it. A search engine will give you a link though.

**Comment by maschlr** [comment]
> @maschlr Mh, for the TG Desktop Variant, it looks like it&#x27;s available independently of the OS ...

Yeah I tried it according to the docs. After choosing "Test Server" on the desktop app, it wants me to scan a QR with my mobile device (Android). There, I get "invalid token". 

Debug menu on mobile also has no test environment option...

**Comment by harshil21** [comment]
@maschlr log in using phone number. Scroll down on this page - https://github.com/telegramdesktop/tdesktop/wiki/Cheat-Codes for more info on test backend. My personal phone number didn't work for me, but the test one did. 

Also if you have an alternative android Telegram client, like nekogram or exteragram, you can use the test server on them too.
